### PR TITLE
CSHARP-4944: Enable use of native crypto in libmongocrypt bindings

### PR DESCRIPTION
**Pull request by adelinowona**
### Description
Use libmongocrypt native crypto when available.

### What's changing

- Bundle crypto-enabled libmongocrypt on Mac.
- On linux continue to bundle crypto-disabled libmongocrypt, but inform users to install the crypto-enabled libmongocrypt on their system and use the new environment variable `LIBMONGOCRYPT_PATH` to make it available to the driver. If `LIBMONGOCRYPT_PATH` is not set then we default to using the packaged crypto-disabled libmongocrypt.

#### Using crypto-disabled libmongocrypt:
| Method                     | ThreadsCount | Mean        | Error     | StdDev    | Median      |
|--------------------------- |------------- |------------:|----------:|----------:|------------:|
| BulkDecryptionUsingBinding | 1            |    50.83 ms |  0.975 ms |  2.279 ms |    49.70 ms |
| BulkDecryptionUsingBinding | 2            |    71.34 ms |  1.422 ms |  3.090 ms |    71.32 ms |
| BulkDecryptionUsingBinding | 8            |   766.59 ms | 23.535 ms | 69.393 ms |   776.18 ms |
| BulkDecryptionUsingBinding | 64           | 5,575.51 ms | 91.840 ms | 85.907 ms | 5,581.51 ms |

#### Using crypto-enabled libmongocrypt:
| Method                     | ThreadsCount | Mean      | Error    | StdDev   |
|--------------------------- |------------- |----------:|---------:|---------:|
| BulkDecryptionUsingBinding | 1            |  22.03 ms | 0.103 ms | 0.096 ms |
| BulkDecryptionUsingBinding | 2            |  22.89 ms | 0.117 ms | 0.104 ms |
| BulkDecryptionUsingBinding | 8            |  30.69 ms | 0.564 ms | 0.554 ms |
| BulkDecryptionUsingBinding | 64           | 201.37 ms | 1.997 ms | 1.559 ms |

### CODE REVIEW

**Review by BorisDog — CHANGES_REQUESTED** [review]
Few minor questions.

File: bindings/cs/MongoDB.Libmongocrypt/LibraryLoader.cs [hunk 1]
```diff
@@ -186,10 +168,18 @@ private class LinuxLibrary : ISharedLibraryLoader
             // #define RTLD_GLOBAL     0x100
             public const int RTLD_GLOBAL = 0x100;
             public const int RTLD_NOW = 0x2;
+            
+            private static readonly string[] _suffixPaths =
```
[the remark this answers is not in the document]
Comment (adelinowona) [reply bindings/cs/MongoDB.Libmongocrypt/LibraryLoader.cs]: Done

File: bindings/cs/README.md [hunk 2]
```diff
@@ -54,6 +54,9 @@ If you see `Windows Error: 126` during tests, like the example below, it means t
 2. cd <build>/bindings/cs
 3. dotnet build cs.build
 ```
+*Note*: You can use the ```LIBMONGOCRYPT_PATH``` environment variable to load a locally installed
```
[the remark this answers is not in the document]
Comment (adelinowona) [reply bindings/cs/README.md:57]: Done

File: bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs [hunk 3]
```diff
@@ -42,14 +42,23 @@ public static CryptClient Create(CryptOptions options)
...
             Status status = null;
 
+            bool cryptoAvailable;
+            try
+            {
+                cryptoAvailable = Library.mongocrypt_is_crypto_available();
+            }
+            catch (LibraryLoader.FunctionNotFoundException)
+            {
+                // mongocrypt_is_crypto_available is only available in libmongocrypt version >= 1.9
```
[the remark this answers is not in the document]
Comment (adelinowona) [reply bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs]: Good point! Done

**Review by BorisDog — CHANGES_REQUESTED** [review]
minor comments

File: bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs [hunk 4]
```diff
@@ -42,14 +42,15 @@ public static CryptClient Create(CryptOptions options)
             MongoCryptSafeHandle handle = null;
             Status status = null;
 
+            // mongocrypt_is_crypto_available is only available in libmongocrypt version >= 1.9
+            var cryptoAvailable = Version.Parse(Library.Version) >= Version.Parse("1.9") && Library.mongocrypt_is_crypto_available();
```
[the remark this answers is not in the document]
Comment (adelinowona) [reply bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs:46]: Done

File: bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs [hunk 5]
```diff
@@ -34,6 +34,9 @@ public class CryptClientFactory
         private static Library.Delegates.RandomCallback __randomCallback = new Library.Delegates.RandomCallback(SecureRandomCallback.GenerateRandom);
         private static Library.Delegates.CryptoHmacCallback __signRsaesPkcs1HmacCallback = new Library.Delegates.CryptoHmacCallback(SigningRSAESPKCSCallback.RsaSign);
 
+        // mongocrypt_is_crypto_available is only available in libmongocrypt version >= 1.9
+        private static readonly Version __s_libVersionWithMongocryptIsCryptoAvailable = Version.Parse("1.9");
```
Comment (BorisDog) [inline bindings/cs/MongoDB.Libmongocrypt/CryptClientFactory.cs:38]: Consider:
`private static readonly Version __mongocryptIsCryptoAvailableMinVersion = Version.Parse("1.9");`

File: bindings/cs/README.md [hunk 6]
```diff
@@ -54,6 +54,9 @@ If you see `Windows Error: 126` during tests, like the example below, it means t
...
+libmongocrypt build. You should specify the absolute path to the libmongocrypt library itself, not just the containing folder.  For example on Linux:
```
Comment (BorisDog) [inline bindings/cs/README.md:58]: I think this comment got lost:
https://github.com/mongodb/libmongocrypt/pull/796#discussion_r1588510088

File: bindings/cs/MongoDB.Libmongocrypt/LibraryLoader.cs [hunk 7]
```diff
@@ -150,9 +124,17 @@ private class DarwinLibraryLoader : ISharedLibraryLoader
             public const int RTLD_GLOBAL = 0x8;
             public const int RTLD_NOW = 0x2;
 
+            private static readonly string[] __s_suffixPaths =
```
Comment (BorisDog) [inline bindings/cs/MongoDB.Libmongocrypt/LibraryLoader.cs:127]: I am sorry for confusion, I wanted to write  `__suffixPaths`  instead  `___s_suffixPaths `

**Review by BorisDog — CHANGES_REQUESTED** [review]

**Review by BorisDog — APPROVED** [review]
LGTM

**Pull request merged**